doc/rm/device_interface_functions.md

```diff
@@ -0,0 +1,166 @@
+DIFs **should not** track hardware-backed state in their own state machine.
+
+A DIF instance is expressly an instance of a hardware block, so DIFs **must not** intrinsically know the location of the hardware blocks they’re associated with in memory.
+That is:
+
+```c
+dif_timer_init_result_t dif_timer_init(dif_timer_t *timer,
+                                       const timer_config_t *config,
+                                       uintptr_t base_address);
+```
+
+allows the system initialization code to instantiate timers at specific locations, whereas:
+
+```c
+// Don't do this:
+bool dif_timer_init(timer_t *timer, enum timer_num num);
+```
+
+suggests that the timer DIF knows about the number and placement of timers in the system.
+
+DIFs **must not** store or provide information in their implementation or state that is outside of their area of concern.
+The number of timers in a system is the concern of the system, not the timer DIF.
+
+## Naming
+
+All DIF functions **must** have clear direct objects and verbs, be written in the imperative mood, and follow the format of `dif_<object>_<verb_phrase>`.
+The object in the name must be common to the DIF it appears it, and unique among DIFs.
+
+Consider the following examples of good names:
+
+*   `dif_timer_init`,
+*   `dif_timer_reset`,
+*   `dif_timer_set_time_remaining`,
+*   `dif_timer_get_time_remaining`.
+
+The following are bad names:
+
+*   `dif_clear_timer` (wrong object/verb-phrase ordering),
+*   `dif_timer_gets_reset` (passive voice not imperative),
+*   `timer_init` (not correctly prefixed for C).
+
+Prefer common names: `init` is more commonly written than `initialize`, but `reset` is more commonly written than `rst`, for example.
+This is a subjective call on the DIF author and reviewers' parts, so common agreement will be more useful than strict prescription.
+
+## Documentation
+
+All DIF exported types and functions **must** have associated API documentation.
+All function parameters **must** be documented.
+All function semantics **must** be documented, no matter how obvious one suspects the function to be.
+The documentation should be exhaustive enough that the implementation can reasonably be inferred from it and the IP specification.
+
+It is important that the DIFs are documented alongside the hardware IP that they are written for.
+Each hardware IP block also contains a "programmers' guide" section, which tells programmers how to use the hardware interfaces correctly, using prose descriptions and relevant diagrams.
+The prose descriptions should include references to relevant DIF functions.
+
+Programmers' guides **should** primarily refer to the [Software API Documentation]({{< relref "/sw#opentitan-software-api-documentation" >}}).
+Programmers' guides **should not** standalone examples that do not match how we implement DIFs and can become out of date.
```

Comment:
"**should not**" `be` "standalone examples" perhaps?

Reply:
ah yes I have missed a word - i think 'contain' is the right word? This is the diff I propose.

```suggestion
Programmers' guides **should not** contain standalone examples that do not match how we implement DIFs and can become out of date.
```